### PR TITLE
[DEVELOPER-2942] Merging in my work for a new docker image

### PR DESCRIPTION
**Pull request by LightGuard**
I've tried drupal-dev which all works correctly.

### CODE REVIEW

**Comment by LightGuard** [comment]
@robpblake This merges our two works together, I wasn't able to get the settings stuff in yet, it took me all day (it's now 21:15) to merge things and iron out all the little bugs.

**Comment by LightGuard** [comment]
Basic login page error.

**Comment by LightGuard** [comment]
retest this please

File: _docker/drupal/scripts/drupal_install_checker.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  def installed?
+    settings_exists? && rhd_settings_exists? && mysql_connect? && tables_exists?
+  end
+
+  def install_drupal
+    puts 'Creating new settings.php file'
+    FileUtils.rm '/var/www/drupal/web/sites/default/settings.php'
+    FileUtils.cp '/var/www/drupal/web/sites/default/default.settings.php','/var/www/drupal/web/sites/default/settings.php'
+    puts 'Installing Drupal, please wait...'
+    puts process_executor.exec!('/var/www/drupal/vendor/bin/drupal', ['site:install', 'standard', '--langcode=en', '--db-type=mysql', '--db-host=drupalmysql', '--db-name=drupal', "--db-user=#{ENV['DB_USER']}", "--db-pass=#{ENV['DB_PASSWORD']}", '--account-name=admin', "--site-name='Red Hat Developers'", "--site-mail='test@example.com'", "--account-mail='admin@example.com'", '--account-pass=admin', '-n'])
```
Comment (robpblake) [inline _docker/drupal/scripts/drupal_install_checker.rb]: In environments like production and staging, the database host/username/password won't be set as environment variables in the container. We also need to consider that the database host is not managed by Docker in production/staging, so I think all of this information needs to be configurable on a per-environment basis.

If this class was able to take a property hash as a constructor parameter, and then this hash was populated from an properties/.yml file at an expected location on the container file system, then that should give us the ability to parameterize on a per-environment basis.

File: _docker/drupal/scripts/drupal_install_checker.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    File.exists? File.join drupal_site, 'settings.php'
+  end
+
+  def rhd_settings_exists?
+    File.exists? File.join drupal_site, 'rhd.settings.php'
+  end
+
+  def mysql_connect?
+    begin
+      process_executor.exec! 'mysql','drupal'
```
Comment (robpblake) [inline _docker/drupal/scripts/drupal_install_checker.rb]: Same comment as below about this needing to be configurable on a per-environment basis.

File: _docker/drupal/db-check.php [hunk 3]
```diff
@@ -10,8 +10,8 @@
  */
 try {
     $DBH = new PDO(
-        "mysql:host=drupalmysql;dbname=$_ENV[DB_NAME]",
```
Comment (robpblake) [inline _docker/drupal/db-check.php]: Same as comments below: We'll need to be able to parameterise the db username/password/host details. My initial thought is that its easiest to move this functionality into your install Ruby script so that the loading of configuration is only done once. Before running the install process, you could check/wait on the database to become available.

**Comment by robpblake** [comment]
@LightGuard This looks good!

I've made a few comments about us needing to parameterise the database details in the scripts within the container. 

Don't know how you want to proceed? Do you want to get this merged and address ^ in another PR or continue working on this one?

cheers,

Rob

**Comment by LightGuard** [comment]
Yeah, I can make those changes. I had originally thought the client setting in the mysql/mariadb config would work for all clients, apparently it doesn't for php :( By the time I got the fix going I forgot about staging/prod needing to be parameterized.

**Comment by LightGuard** [comment]
@robpblake take a look now, kick it around a bit and see if this will work for you.

This relates to https://issues.jboss.org/browse/DEVELOPER-2942

**Comment by LightGuard** [comment]
It also enables https://issues.jboss.org/browse/DEVELOPER-2952

**Comment by robpblake** [comment]
@LightGuard Looks like there might be a small issue somewhere. When I try and run:

`bundle exec ./control.rb -e drupal-dev --run-the-stack`

Drupal does not start and I get this error on the command line of the Drupal  container:

```
/usr/local/lib/ruby/2.1.0/open3.rb:193:in `spawn': No such file or directory - /var/www/drupal/vendor/bin/drupal (Errno::ENOENT)
	from /usr/local/lib/ruby/2.1.0/open3.rb:193:in `popen_run'
	from /usr/local/lib/ruby/2.1.0/open3.rb:188:in `popen2e'
	from /usr/local/lib/ruby/2.1.0/open3.rb:327:in `capture2e'
	from ../scripts/drupal_install_checker.rb:8:in `exec!'
	from ../scripts/drupal_install_checker.rb:71:in `install_drupal'
	from ../scripts/drupal_install_checker.rb:97:in `<main>'
```
I'm not sure how `/var/www/drupal/vendor/bin/drupal` is being installed, so couldn't immediately determine how to resolve it.

**Comment by robpblake** [comment]
@LightGuard I'm also see this git status on the branch after having run `--run-the-stack`:

```
Robs-Mac-mini:_docker Rob$ git status
On branch developer-1122
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   drupal/web/sites/default/settings.php

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	drupal/themes/

no changes added to commit (use "git add" and/or "git commit -a")
Robs-Mac-mini:_docker Rob$
```

**Comment by LightGuard** [comment]
Oops, yeah, we need to run composer install first thing, forgot about that.

On Thursday, May 26, 2016, Rob Blake <notifications@github.com> wrote:

> @LightGuard <https://github.com/LightGuard> Looks like there might be a
> small issue somewhere. When I try and run:
>
> bundle exec ./control.rb -e drupal-dev --run-the-stack
>
> Drupal does not start and I get this error on the command line of the
> Drupal container:
>
> /usr/local/lib/ruby/2.1.0/open3.rb:193:in `spawn': No such file or directory - /var/www/drupal/vendor/bin/drupal (Errno::ENOENT)
>     from /usr/local/lib/ruby/2.1.0/open3.rb:193:in `popen_run'
>     from /usr/local/lib/ruby/2.1.0/open3.rb:188:in `popen2e'
>     from /usr/local/lib/ruby/2.1.0/open3.rb:327:in `capture2e'
>     from ../scripts/drupal_install_checker.rb:8:in `exec!'
>     from ../scripts/drupal_install_checker.rb:71:in `install_drupal'
>     from ../scripts/drupal_install_checker.rb:97:in `<main>'
>
> I'm not sure how /var/www/drupal/vendor/bin/drupal is being installed, so
> couldn't immediately determine how to resolve it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/redhat-developer/developers.redhat.com/pull/1122#issuecomment-221827190>
>


-- 
Sent from Gmail Mobile

**Comment by LightGuard** [comment]
I thought I ignored that one :( it's going to change all the time for dev
and pr. I figured for staging and prod we'd put everything we need into
the... Crap that won't work. Gotta talk this one through a bit I think.

On Thursday, May 26, 2016, Rob Blake <notifications@github.com> wrote:

> @LightGuard <https://github.com/LightGuard> I'm also see this git status
> on the branch after having run --run-the-stack:
>
> Robs-Mac-mini:_docker Rob$ git status
> On branch developer-1122
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git checkout -- <file>..." to discard changes in working directory)
>
>     modified:   drupal/web/sites/default/settings.php
>
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
>
>     drupal/themes/
>
> no changes added to commit (use "git add" and/or "git commit -a")
> Robs-Mac-mini:_docker Rob$
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/redhat-developer/developers.redhat.com/pull/1122#issuecomment-221841309>
>


-- 
Sent from Gmail Mobile

**Comment by LightGuard** [comment]
Alright @robpblake take four.

**Comment by robpblake** [comment]
@LightGuard Still a few issues for me on this, but I got further than last time:


###Issue 1: 

First error I received was this from within the drupal container:

```
Robs-Mac-mini:~ Rob$ docker logs drupaldev_drupal_1
/usr/local/lib/ruby/2.1.0/psych.rb:464:in `initialize': No such file or directory @ rb_sysopen - /var/www/drupal/web/sites/default/rhd.settings.yml (Errno::ENOENT)
	from /usr/local/lib/ruby/2.1.0/psych.rb:464:in `open'
	from /usr/local/lib/ruby/2.1.0/psych.rb:464:in `load_file'
	from scripts/drupal_install_checker.rb:107:in `<main>'
```

This was being caused by the rhd.settings.yml.erb not being templated for any other environment other than drupal-pull-request.

I resolved this by doing the following:

1. Re-named `_docker/environments/drupal-dev/rhd.settings.yml.erb` -> `_docker/environments/drupal-dev/rhd.settings.yml`
2: Added the following bind mount to the drupal service definition in `_docker/environments/drupal-dev/docker-compose.yml`:

```
volumes:
      - ./rhd.settings.yml:/var/www/drupal/web/sites/default/rhd.settings.yml
``` 

### 2: Git status after `--run-the-stack`

I'm still seeing changes to my local workspace after simply checking out and then running `--run-the-stack`:

```
On branch developer-1122
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
	
        modified:   _docker/drupal/web/sites/default/settings.php

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	_docker/drupal/web/profiles/
```

### 3: Very slow build and container death

On my machine I cannot get the `--run-the-stack` to complete. It takes about ~15 minutes to start the container and then eventually it starts to run awestruct. However a few minutes into the awestruct process when it starts to push content into Drupal I get:

```
Generating pages...
Using the following options from site.generation: {:in_theads=>20, :in_processes=>2}
Error making drupal request to '/about/index.html', retrying...
Error making drupal request to '/articles/no-cost-rhel-faq/index.html', retrying...
Error making drupal request to '/about/index.html', retrying...
Error making drupal request to '/articles/no-cost-rhel-faq/index.html', retrying...
Error making drupal request to '/about/index.html', retrying...
Error making drupal request to '/articles/no-cost-rhel-faq/index.html', retrying...
Error making drupal request to '/about/index.html', retrying...
Error making drupal request to '/articles/no-cost-rhel-faq/index.html', retrying...
Error making drupal request to '/about/index.html', retrying...
```

And then the container dies and I have to completely re-start docker. I suspect this is something local to my machine, but I haven't been able to work out what yet. However it would be good to get someone else to try it on their machine to make sure I'm not the only one who experiences this.

**Comment by LightGuard** [comment]
That looks like something happened where Drupal didn't startup correctly.

On Friday, May 27, 2016, Rob Blake <notifications@github.com> wrote:

> @LightGuard <https://github.com/LightGuard> Still a few issues for me on
> this, but I got further than last time:
>
> ###Issue 1:
>
> First error I received was this from within the drupal container:
>
> Robs-Mac-mini:~ Rob$ docker logs drupaldev_drupal_1
> /usr/local/lib/ruby/2.1.0/psych.rb:464:in `initialize': No such file or directory @ rb_sysopen - /var/www/drupal/web/sites/default/rhd.settings.yml (Errno::ENOENT)
>     from /usr/local/lib/ruby/2.1.0/psych.rb:464:in `open'
>     from /usr/local/lib/ruby/2.1.0/psych.rb:464:in `load_file'
>     from scripts/drupal_install_checker.rb:107:in `<main>'
>
> This was being caused by the rhd.settings.yml.erb not being templated for
> any other environment other than drupal-pull-request.
>
> I resolved this by doing the following:
>
>    1. Re-named _docker/environments/drupal-dev/rhd.settings.yml.erb ->
>    _docker/environments/drupal-dev/rhd.settings.yml 2: Added the
>    following bind mount to the drupal service definition in
>    _docker/environments/drupal-dev/docker-compose.yml:
>
> volumes:
>       - ./rhd.settings.yml:/var/www/drupal/web/sites/default/rhd.settings.yml
>
> 2: Git status after --run-the-stack
>
> I'm still seeing changes to my local workspace after simply checking out
> and then running --run-the-stack:
>
> On branch developer-1122
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git checkout -- <file>..." to discard changes in working directory)
>
>         modified:   _docker/drupal/web/sites/default/settings.php
>
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
>
>     _docker/drupal/web/profiles/
>
> 3: Very slow build and container death
>
> On my machine I cannot get the --run-the-stack to complete. It takes
> about ~15 minutes to start the container and then eventually it starts to
> run awestruct. However a few minutes into the awestruct process when it
> starts to push content into Drupal I get:
>
> Generating pages...
> Using the following options from site.generation: {:in_theads=>20, :in_processes=>2}
> Error making drupal request to '/about/index.html', retrying...
> Error making drupal request to '/articles/no-cost-rhel-faq/index.html', retrying...
> Error making drupal request to '/about/index.html', retrying...
> Error making drupal request to '/articles/no-cost-rhel-faq/index.html', retrying...
> Error making drupal request to '/about/index.html', retrying...
> Error making drupal request to '/articles/no-cost-rhel-faq/index.html', retrying...
> Error making drupal request to '/about/index.html', retrying...
> Error making drupal request to '/articles/no-cost-rhel-faq/index.html', retrying...
> Error making drupal request to '/about/index.html', retrying...
>
> And then the container dies and I have to completely re-start docker. I
> suspect this is something local to my machine, but I haven't been able to
> work out what yet. However it would be good to get someone else to try it
> on their machine to make sure I'm not the only one who experiences this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/redhat-developer/developers.redhat.com/pull/1122#issuecomment-222117795>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAB1EzEnMvEKTviF-K7LkURlXdIyicv_ks5qFsy8gaJpZM4IlEji>
> .
>


-- 
Sent from Gmail Mobile

**Comment by LightGuard** [comment]
Also the startup time on the first run is going to take a while since you have to fetch everything via composer for the drupal build.

**Comment by robpblake** [comment]
@LightGuard 

I'm still having issues, but I've managed to dig a bit further into it.

### 1: Bind mount permissions

One of the issues I believe is down to file system permissions of the bind mount on Macs. When we bind mount the _docker/drupal directory into the container, it ends up being owned by user id '1000' and group 'staff'. This is my userid and group on Mac OS. This causes a problem because the Apache/Drupal process is running with user apache (id 24), which therefore does not have access to the bind mount volumes for writing (which it apparently needs).

I managed to make a bit of progress here by adding the following to the Dockerfile for the Drupal image:

```
RUN groupadd staff

RUN usermod -u 1000 apache
RUN usermod -G staff apache
```

Clearly this will only work for Mac users, so I'm wondering if the best work-around here is to simply make apache within the container run as root? That should mean that it has all permissions to any directory that it needs.

### 2: 500 internal server error response from Drupal when POST-ing content

I'm getting a 500 internal server error from Drupal when we try to post content to it. The drupal logs in the database seem to indicate this problem:

```
                                                 |        5 |      | http://drupal/entity/node               |         | 172.26.0.6 | 1464698609 |
| 166 |   1 | php            | %type: @message in %function (line %line of %file). | a:5:{s:5:"%type";s:14:"LogicException";s:8:"@message";s:211:"The controller result claims to be providing relevant cache metadata, but leaked metadata was detected. Please ensure you are not rendering content too early. Returned object class: Drupal\rest\ResourceResponse.";s:9:"%function";s:111:"Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext()";s:5:"%file";s:102:"/var/www/drupal/web/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php";s:5:"%line";i:154;} |        3 |      | http://drupal/entity/node     
```

Pretty cryptic, but [this post](http://drupal.stackexchange.com/questions/196520/creating-node-via-rest-post-gives-leaked-cache-metadata-error) seems to suggest that [this is the underlying problem](https://www.drupal.org/node/2626298), which is apparently fixed in Drupal 8.2. 

Have you managed to get a full `control.rb -e drupal-dev --run-the-stack` to work successfully?

**Comment by robpblake** [comment]
@LightGuard I'm also still seeing the following after `git status`:

```
obs-Mac-mini:drupal Rob$ git status
On branch developer-1122-drupal-build
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	web/profiles/

no changes added to commit (use "git add" and/or "git commit -a")
```

Think we're going to need `web/profiles` added to .gitignore.

**Comment by LightGuard** [comment]
@robpblake Okay, give this one a try, I hope I got it all this time. *crosses fingers*

**Comment by Dantheman720** [comment]
@robpblake @LightGuard This is unrelated to drupal builds, but it relates to some problems I've been having with my mac VM build. I've tried removing all docker images and processes and building from scratch.  My results are as follows:
[ErrorWoes.txt](https://github.com/redhat-developer/developers.redhat.com/files/292369/ErrorWoes.txt)

**Comment by robpblake** [comment]
@Dantheman720 Assuming you haven't already resolved your problems, give me a shout on HipChat when you're online.

cheers,

Rob

**Comment by robpblake** [comment]
@LightGuard Progress again, but unfortunately I can't get this running on my machine.

If I run `bundle exec ./control.rb -e drupal-dev --run-the-stack` I get all the way through to the point where it starts to push content into Drupal and after a couple of minutes, the Docker daemon on my host dies.

At that point I have to re-start Docker completely meaning that I've lost all progress. I'm trying it on another machine now, but it would be good to see if other people can get this working.

As to a resolution to the above, I'm really not sure.

**Comment by LightGuard** [comment]
It dies? No messages or errors? Does it need more memory?

On Wednesday, June 1, 2016, Rob Blake <notifications@github.com> wrote:

> @LightGuard <https://github.com/LightGuard> Progress again, but
> unfortunately I can't get this running on my machine.
>
> If I run bundle exec ./control.rb -e drupal-dev --run-the-stack I get all
> the way through to the point where it starts to push content into Drupal
> and after a couple of minutes, the Docker daemon on my host dies.
>
> At that point I have to re-start Docker completely meaning that I've lost
> all progress. I'm trying it on another machine now, but it would be good to
> see if other people can get this working.
>
> As to a resolution to the above, I'm really not sure.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/redhat-developer/developers.redhat.com/pull/1122#issuecomment-222974299>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAB1E0rbL-Y4CEeW_XW-Fkt9CMJFsfAzks5qHXelgaJpZM4IlEji>
> .
>


-- 
Sent from Gmail Mobile

**Comment by robpblake** [comment]
@LightGuard No messages at all and nothing in the docker daemon log file.

The VM has 12G of RAM and access to all of my CPUs, which I would hope would be enough.

I'm trying it on another machine now

**Comment by LightGuard** [comment]
@robpblake https://github.com/docker/docker/issues/15905 says check dmesg
to see if the seam on died via oom or some other means.

On Wednesday, June 1, 2016, Rob Blake <notifications@github.com> wrote:

> @LightGuard <https://github.com/LightGuard> No messages at all and
> nothing in the docker daemon log file.
>
> The VM has 12G of RAM and access to all of my CPUs, which I would hope
> would be enough.
>
> I'm trying it on another machine now
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/redhat-developer/developers.redhat.com/pull/1122#issuecomment-222990566>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAB1E3f0ck6IP5qI9evMl_OBA2gK_aBJks5qHYfWgaJpZM4IlEji>
> .
>


-- 
Sent from Gmail Mobile

**Comment by LightGuard** [comment]
At this point should I go back to an Ubuntu image?

On Wednesday, June 1, 2016, Jason Porter <lightguard.jp@gmail.com> wrote:

> @robpblake https://github.com/docker/docker/issues/15905 says check dmesg
> to see if the seam on died via oom or some other means.
>
> On Wednesday, June 1, 2016, Rob Blake <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> @LightGuard <https://github.com/LightGuard> No messages at all and
>> nothing in the docker daemon log file.
>>
>> The VM has 12G of RAM and access to all of my CPUs, which I would hope
>> would be enough.
>>
>> I'm trying it on another machine now
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/redhat-developer/developers.redhat.com/pull/1122#issuecomment-222990566>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AAB1E3f0ck6IP5qI9evMl_OBA2gK_aBJks5qHYfWgaJpZM4IlEji>
>> .
>>
>
>
> --
> Sent from Gmail Mobile
>


-- 
Sent from Gmail Mobile

**Comment by robpblake** [comment]
@LightGuard : @paulrobinson managed to get a build on his machine. It was very slow, but we got a build.

I'm just updating "all the things" on my mac to ensure I'm running the latest and greatest versions of docker, as Paul did this yesterday.

cheers,

Rob

**Comment by LightGuard** [comment]
Is a subsequent build any faster?

On Wednesday, June 1, 2016, Rob Blake <notifications@github.com> wrote:

> @LightGuard <https://github.com/LightGuard> : @paulrobinson
> <https://github.com/paulrobinson> managed to get a build on his machine.
> It was very slow, but we got a build.
>
> I'm just updating "all the things" on my mac to ensure I'm running the
> latest and greatest versions of docker, as Paul did this yesterday.
>
> cheers,
>
> Rob
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/redhat-developer/developers.redhat.com/pull/1122#issuecomment-223009042>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAB1E561LKoiBm0wAhbQ9pxVTeUfAnwsks5qHZZxgaJpZM4IlEji>
> .
>


-- 
Sent from Gmail Mobile

**Comment by robpblake** [comment]
@LightGuard I was able to get a build on my laptop (so the docker daemon dying problem is related to my desktop). It did however take about 1 hr from start to finish.

I'll try it again and see if there is any improvement in speed 2nd time around.

**Comment by robpblake** [comment]
Stats for a re-run on my laptop:

```
Daybreak database _tmp/cache.db was not closed, state might be inconsistent

real	37m53.094s
user	3m13.366s
sys	1m17.734s
Robs-Air:_docker Rob$ 
```

That is a VM with 6G of memory and 1 CPU (I think i5)

**Comment by robpblake** [comment]
Stats for a re-run on my laptop:

```
Daybreak database _tmp/cache.db was not closed, state might be inconsistent

real	37m53.094s
user	3m13.366s
sys	1m17.734s
Robs-Air:_docker Rob$ 
```

That is a VM with 6G of memory and 1 CPU (I think i5)

**Comment by robpblake** [comment]
Stats for a re-run on my laptop:

```
Daybreak database _tmp/cache.db was not closed, state might be inconsistent

real	37m53.094s
user	3m13.366s
sys	1m17.734s
Robs-Air:_docker Rob$ 
```

That is a VM with 6G of memory and 1 CPU (I think i5)

**Comment by LightGuard** [comment]
@robpblake 35 minutes?? ugh. Where is it spending the most time?

**Comment by robpblake** [comment]
@LightGuard I would say ~10 mins installing Drupal, the other time is spent on Awestruct

**Comment by LightGuard** [comment]
Okay, well with 1 CPU, yeah awestruct is going to take a while.

**Comment by LightGuard** [comment]
@robpblake Can you give it another CPU and see how things change? Was @paulrobinson able to get it to work?

**Comment by robpblake** [comment]
@LightGuard You might like to see this:

```
Daybreak database _tmp/cache.db was not closed, state might be inconsistent

real	12m51.545s
user	0m49.714s
sys	0m19.183s
```
That is a complete `--run-the-stack` on my desktop (12G with 2 cores). So, still not amazingly quick but way better than the ~40 minutes I was seeing yesterday.

It looks like the problem on Macs is down to the filesystem performance being absolutely pants when you're using it as a bind mount with virtualbox. I tried it with the Docker for Mac beta, and it was actually worse than straight-up virtualbox so changing to the beta doesn't really help at the minute (hopefully they can resolve the issue before beta goes GA).

As a work-around I mounted the required volumes into the docker-machine vm using NFS (I found this really useful project to help: https://github.com/adlogix/docker-machine-nfs). With that done, I can get a repeatable, stable build with the above times.

So in conclusion, for mac users there is another step to work-around, but it now seems do-able.

**Comment by robpblake** [comment]
@LightGuard A couple of other things that I noticed whilst working on this today:

### 1: Permissions hack can be removed from `drupal/Dockerfile`:

The following is no longer needed and can be removed:

```
# Hack for permissions
RUN groupadd staff && \
    usermod -u 1000 apache && \
    usermod -G staff apache
```

### 2: Database name needs to be parameterisable in `drupal_install_checker.rb`

In the same way that you're using `--db-host=#{@opts['database']['host']}`, the database name also needs to be parameterisable as in staging and production, the database name given to us by IT is `rhd_mysql`

### 3: Update of twig files to reference properties loaded from `rhd.settings.yml`

Are you able to have a go at updating the twig files to reference the properties loaded from rhd.settings.yml? E.g. `app.baseUrl = "http://localhost.localdomain:32769";` in node--article.html.twig

**Comment by LightGuard** [comment]
One and two I can take care of here no problem. Three I think I'd rather do
in another PR.

On Thursday, June 2, 2016, Rob Blake <notifications@github.com> wrote:

> @LightGuard <https://github.com/LightGuard> A couple of other things that
> I noticed whilst working on this today:
> 1: Permissions hack can be removed from drupal/Dockerfile:
>
> The following is no longer needed and can be removed:
>
> # Hack for permissions
> RUN groupadd staff && \
>     usermod -u 1000 apache && \
>     usermod -G staff apache
>
> 2: Database name needs to be parameterisable in drupal_install_checker.rb
>
> In the same way that you're using --db-host=#{@opts['database']['host']},
> the database name also needs to be parameterisable as in staging and
> production, the database name given to us by IT is rhd_mysql
> 3: Update of twig files to reference properties loaded from
> rhd.settings.yml
>
> Are you able to have a go at updating the twig files to reference the
> properties loaded from rhd.settings.yml? E.g. app.baseUrl = "
> http://localhost.localdomain:32769"; in node--article.html.twig
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/redhat-developer/developers.redhat.com/pull/1122#issuecomment-223245435>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAB1ExqODEmnmm-O-yIz9MWUcMrQGWC_ks5qHqabgaJpZM4IlEji>
> .
>


-- 
Sent from Gmail Mobile

**Comment by robpblake** [comment]
@LightGuard Sounds good.
